x/rollapp/keeper/block_height_to_finalization_queue.go

```diff
@@ -141,26 +156,80 @@ func (k *Keeper) finalizePendingState(ctx sdk.Context, stateInfoIndex types.Stat
 	return nil
 }
 
-func (k Keeper) updateQueueForHeight(ctx sdk.Context, blockHeightToFinalizationQueue types.BlockHeightToFinalizationQueue, failedRollapps map[string]uint64) {
-	// remove the blockHeightToFinalizationQueue if all the rollapps' states are finalized
-	if len(failedRollapps) == 0 {
-		k.RemoveBlockHeightToFinalizationQueue(ctx, blockHeightToFinalizationQueue.CreationHeight)
-		return
+// SetFinalizationQueue set types.BlockHeightToFinalizationQueue for a specific height and rollappID.
+// Panics on encoding errors.
+func (k Keeper) SetFinalizationQueue(ctx sdk.Context, queue types.BlockHeightToFinalizationQueue) {
+	err := k.finalizationQueue.Set(ctx, collections.Join(queue.CreationHeight, queue.RollappId), queue)
+	if err != nil {
+		panic(err)
 	}
-	// remove from the queue only the rollapps that were successfully finalized at all indices.
-	// while iterating the queue for deleting the successfully finalized states, we remove them if
-	// - rollapp was not found in the failedRollapps map
-	// - if it was found, the indexes to be deleted should only be the ones up to the point of the index of the first failed state change
-	blockHeightToFinalizationQueue.FinalizationQueue = slices.DeleteFunc(blockHeightToFinalizationQueue.FinalizationQueue,
-		func(si types.StateInfoIndex) bool {
-			idx, failed := failedRollapps[si.RollappId]
-			return !failed || si.Index < idx
-		})
-	// save the current queue with "leftover" rollapp's state changes
-	k.SetBlockHeightToFinalizationQueue(ctx, blockHeightToFinalizationQueue)
+}
+
+// GetFinalizationQueue gets types.BlockHeightToFinalizationQueue for a specific height and rollappID.
+// Panics on encoding errors.
+func (k Keeper) GetFinalizationQueue(ctx sdk.Context, height uint64, rollappID string) (types.BlockHeightToFinalizationQueue, bool) {
+	queue, err := k.finalizationQueue.Get(ctx, collections.Join(height, rollappID))
+	if err != nil && !errors.Is(err, collections.ErrNotFound) {
+		panic(err)
+	}
+	found := err == nil
+	return queue, found
+}
+
+// RemoveFinalizationQueue removes types.BlockHeightToFinalizationQueue.
+// Panics on encoding errors. Do not panic if the key does not exist.
```

Comment:
seems like we panic here also if not found? don't see where we check the error.

Reply:
we don't need an additional check, collections api guarantees that "not found" doesn't panic (i double-checked!)
https://github.com/cosmos/cosmos-sdk/blob/main/collections/map.go#L131